mava/systems/q_learning/sebulba/rec_iql.py

```diff
@@ -0,0 +1,782 @@
+        for thread_id in range(config.arch.n_threads_per_executor):
+            key, act_key = jax.random.split(key)
+            seeds = np_rng.integers(np.iinfo(np.int32).max, size=config.arch.num_envs).tolist()
+            act_key = jax.device_put(key, actor_device)
+            actor_id = device_idx * config.arch.n_threads_per_executor + thread_id
+
+            actor = threading.Thread(
+                target=rollout,
+                args=(
+                    act_key,
+                    # We have to do this here, creating envs inside actor threads causes deadlocks
+                    environments.make_gym_env(config, config.arch.num_envs),
+                    config,
+                    pipe,
+                    params_source,
+                    q_net,
+                    actor_device,
+                    seeds,
+                    actor_lifetime,
+                    actor_id,
+                ),
+                name=f"Actor-{actor_device}-{thread_id}",
+            )
+            actor_threads.append(actor)
+
+    # Start the actors simultaneously
+    for actor in actor_threads:
+        actor.start()
+
+    eval_queue: Queue = Queue()
+    threading.Thread(
+        target=learner_thread,
+        name="Learner",
+        args=(learn, learner_state, config, eval_queue, pipe, params_sources),
+    ).start()
+
+    max_episode_return = -np.inf
+    best_params_cpu = jax.device_get(inital_params.online)
+
+    eval_hs = ScannedRNN.initialize_carry(
+        (min(config.arch.num_eval_episodes, config.arch.num_envs), config.system.num_agents),
+        config.network.hidden_state_dim,
+    )
+
+    # This is the main loop, all it does is evaluation and logging.
+    # Acting and learning is happening in their own threads.
+    # This loop waits for the learner to finish an update before evaluation and logging.
+    for eval_step in range(config.arch.num_evaluation):
+        # Sync with the learner - the get() is blocking so it keeps eval and learning in step.
+        episode_metrics, train_metrics, learner_state, time_metrics = eval_queue.get()
+
+        t = int(steps_per_rollout * (eval_step + 1))
+        time_metrics |= {"timestep": t, "pipline_size": pipe.qsize()}
+        logger.log(time_metrics, t, eval_step, LogEvent.MISC)
+
+        if episode_metrics:
+            episode_metrics, ep_completed = get_final_step_metrics(episode_metrics)
+            episode_metrics["steps_per_second"] = steps_per_rollout / time_metrics["rollout_time"]
+            if ep_completed:
+                logger.log(episode_metrics, t, eval_step, LogEvent.ACT)
+
+            train_metrics["learner_step"] = (eval_step + 1) * config.system.num_updates_per_eval
+            train_metrics["learner_steps_per_second"] = (
+                config.system.num_updates_per_eval
+            ) / time_metrics["learner_time_per_eval"]
+            logger.log(train_metrics, t, eval_step, LogEvent.TRAIN)
```

Comment:
I think we should always log train metrics even if an episode hasn't finished yet, what do you think?